patterns/structural/proxy.py

```diff
@@ -1,55 +1,67 @@
+changing its interface. The main class is called `Real Subject`. A client should
+use the proxy or the real subject without any code change, so both must have the
+same interface. Logging and controlling access to the real subject are some of
+the proxy pattern usages.
+
+*References:
+https://refactoring.guru/design-patterns/proxy/python/example
+https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Fronting.html
+
 *TL;DR
-Provides an interface to resource that is expensive to duplicate.
+Add functionality to a resource without changing its interface.
 """
 
-import time
 
+class Subject:
+    """
+    As mentioned in the document, interfaces of both RealSubject and Proxy should
+    be the same, because the client should be able to use RealSubject or Proxy with
+    no code change.
+    """
+
+    def do_the_job(self, user):
+        raise NotImplementedError()
+
+
+class RealSubject(Subject):
+    """
+    This is the main job doer. External services like payment gateways can be a
+    good example.
+    """
 
-class SalesManager:
-    def talk(self):
-        print("Sales Manager ready to talk")
+    def do_the_job(self, user):
+        print(f'> I am doing the job for {user}')
 
 
-class Proxy:
+class Proxy(Subject):
     def __init__(self):
-        self.busy = 'No'
-        self.sales = None
-
-    def talk(self):
-        print("Proxy checking for Sales Manager availability")
-        if self.busy == 'No':
-            self.sales = SalesManager()
-            time.sleep(0.1)
-            self.sales.talk()
-        else:
-            time.sleep(0.1)
-            print("Sales Manager is busy")
+        self._real_subject = RealSubject()
+
+    def do_the_job(self, user):
+        """
+        logging and controlling access are some examples of proxy usages.
+        """
 
+        print(f'[log] Doing the job for {user} is requested.')
 
-class NoTalkProxy(Proxy):
-    def talk(self):
-        print("Proxy checking for Sales Manager availability")
-        time.sleep(0.1)
-        print("This Sales Manager will not talk to you", "whether he/she is busy or not")
+        if user == 'admin':
+            self._real_subject.do_the_job(user)
+        else:
+            print(f'[log] I can do the job just for `admins`.')
 
 
 if __name__ == '__main__':
-    p = Proxy()
-    p.talk()
-    p.busy = 'Yes'
-    p.talk()
-    p = NoTalkProxy()
-    p.talk()
-    p.busy = 'Yes'
-    p.talk()
-
-### OUTPUT ###
-# Proxy checking for Sales Manager availability
-# Sales Manager ready to talk
-# Proxy checking for Sales Manager availability
-# Sales Manager is busy
-# Proxy checking for Sales Manager availability
-# This Sales Manager will not talk to you whether he/she is busy or not
-# Proxy checking for Sales Manager availability
-# This Sales Manager will not talk to you whether he/she is busy or not
+    proxy = Proxy()
+    proxy.do_the_job('admin')
+    proxy.do_the_job('anonymous')
```

Comment:
Could you please also show here that client can call RealSubject the same way as the Proxy? (the essential "interface was not changed" part of this pattern). Results would be different ofcourse.